### ISSUE TITLE
TLS 1.3 fails to resume PSK sessions

### ISSUE THREAD

**Issue by olszomal** [openssl/openssl#7433]
OpenSSL 1.1.1 (or 1.1.2-dev) is unable to resume TLS 1.3 sessions created with a PSK cipher suite.

### Steps to reproduce the bug

I reproduced this problem using s_server and s_client diagnostic tools.

I configured a PSK server:
```
openssl s_server -accept 4433 -psk 1a2b3c4d -nocert
```

1st `openssl s_client` connection to create and save a session (works correctly):
```
openssl s_client -connect 127.0.0.1:4433 -psk 1a2b3c4d -sess_out sesion
```

2nd `openssl s_client` connection to resume the saved session (fails):
```
openssl s_client -connect 127.0.0.1:4433 -psk 1a2b3c4d -sess_in sesion
CONNECTED(00000005)
140476586415744:error:14094417:SSL routines:ssl3_read_bytes:sslv3 alert illegal parameter:ssl/record/rec_layer_s3.c:1528:SSL alert number 47
```

The printed `openssl s_server` logs:
```
PSK warning: client identity not what we expected (got '<a number of garbage bytes>' expected 'Client_identity')
ERROR
139982316717696:error:141FA0FD:SSL routines:tls_psk_do_binder:binder does not verify:ssl/statem/extensions.c:1604:
```

### My initial (and incomplete) analysis

I diagnosed the garbage bytes to be generated by the `construct_stateless_ticket()` function on the server side, transferred via the `pkt` structure to the client, and stored in the `s->session->ext.tick` ticket on the client side.

On session resumption attempt only, the `tls_construct_ctos_psk()` function stores this value in the `pkt` structure on the client side. This value then becomes a parameter of the `psk_find_session_cb()` callback, producing the `PSK warning` message. Our `s_server` fails a bit further in the `tls_psk_do_binder()` function.

I guess the issue is caused by a discrepancy between the server-side and client-side interpretation of the `pkt` structure.

It seems to be related to #7312

**Comment by mattcaswell** [comment]
That's actually an s_server bug. In TLSv1.3 external PSKs and session resumptions are indistinguishable. The PSK callback needs to fail gracefully if it doesn't recognise the identity. Libssl will continue checking to see if the "PSK" is actually a session resumption.

Fix in #7434

**Comment by mtrojnar** [comment]
I observed the same problem with stunnel, so I don't think this is just s_server.
Also, comparing `identity_len` to the expected value (as implemented in #7434) won't work in real servers, which may accept several PSK identities.  Some of those identities may have the same length as the session resumption.

**Comment by mattcaswell** [comment]
> Also, comparing identity_len to the expected value (as implemented in #7434) won't work in real servers, which may accept several PSK identities. Some of those identities may have the same length as the session resumption.

Any real callback needs to set the session to NULL and return 1 if it can't find a match for the identity. In s_server we only have one possible identity to match, so this way works just fine. The logic might look a little more complicated in a server with multiple valid identities but the principle is the same.

**Comment by mtrojnar** [comment]
@mattcaswell Thank you.
